samples/ffmpeg-wasm-demuxer/download_reader.js

```diff
@@ -0,0 +1,105 @@
+import {SharedReadBuffer} from './shared_read_buffer.js'
+
+class DownloadSink {
+  #onChunk = null;
+  #onEOF = null;
+
+  constructor(onChunk, onEOF) {
+    this.#onChunk = onChunk;
+    this.#onEOF = onEOF;
+  }
+
+  write(chunk) {
+    this.#onChunk(chunk);
+  }
+
+  close() {
+    this.#onEOF();
+  }
+}
+
+export class DownloadReader {
+  #fileStorage = null;
+  #writePtr = 0;
+  #readPtr = 0;
+  #eof = false;
+  #pendingReadNumBytes = 0;
+  #sharedReadBuffer = null;
+  #enableLogging = false;
+
+  async initialize(fileUri, sab) {
+    this.#sharedReadBuffer = new SharedReadBuffer(sab);
+
+    // Fetch the file and pipe the data through.
+    const downloadSink = new DownloadSink(this._onChunk.bind(this), this._onEOF.bind(this));
+    this._log('starting fetch');
+    fetch(fileUri).then(response => {
+      // highWaterMark should be large enough for smooth streaming, but lower is
+      // better for memory usage.
+      response.body.pipeTo(new WritableStream(downloadSink, {highWaterMark: 2}));
+    });
+  }
+
+  read(numBytes) {
+    this._log(`read(${numBytes})`);
+    let bytesUnread = this.#writePtr - this.#readPtr;
+    let cappedForEOF = false;
+    if (numBytes > bytesUnread) {
+      if (!this.#eof) {
+        this._log(`waiting for more chunks`);
+        // Read will complete in _onChunk() once enough bytes become available.
+        this.#pendingReadNumBytes = numBytes;
+        return;
+      } else {
+        this._log(`capping read for EOF`);
+        cappedForEOF = true;
+        numBytes = bytesUnread;
+      }
+    }
+
+    this._log(`read() - fulfilling now!`);
+
+    this.#pendingReadNumBytes = 0;
+    this.#sharedReadBuffer.write(new Uint8Array(this.#fileStorage.buffer,
+        this.#readPtr, numBytes), cappedForEOF);
+    this.#readPtr += numBytes;
+  }
+
+  _onChunk(chunk) {
```

Comment:
Nit: Private methods can also just be `#` (although there may be cases where that is impossible because a "caller" is external; `#` private is very private).